free up dispatch key space - python changes

bdhirsh opened the pull request:
This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69633 free up dispatch key space (in C++)
* **#68744 free up dispatch key space - python changes**
* #68743 Implement a mostly-faithful dispatcher in python

Differential Revision: [D33255194](https://our.internmc.facebook.com/intern/diff/D33255194)

bdhirsh committed:
free up dispatch key space - python changes

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh commented on torch/_mock_dispatcher/dispatch_key_extractor.py:
```diff
@@ -1,21 +1,72 @@
+                return
+        self.requiresBitsetPerBackend_ = False
+
     def getDispatchKeySetUnboxed(self, args: List[DispatchKeySet]) -> DispatchKeySet:
```
I added @bhosmer's optimization here: for ops that don't have backend-dependent fall through keys, we can use a single dispatch keyset instead of having to grab the correct one from a list at runtime. This change:

(-) Adds a (very predictable) branch in the hotpath
(+) Avoids a load from a global array of bitsets, when the correct branch is taken

Right now, the good path will be taken for every aten op, since none of Autograd/Dense/Sparse/Quantized register explicit fallthroughs to specific aten ops. If we make Autocast a per-backend key at some point, then a few ops would end up taking the slower path, since according to @ngimel we register a different set of fall through kernels to AutocastCPU/AutocastCUDA.

zou3519 commented on torch/_mock_dispatcher/dispatch_key.py:
```diff
@@ -1,47 +1,96 @@
+    # Dense lives above the "Other" backends, mostly because the other backends should have higher priority than dense CPU.
+    # e.g. torch.add(vulkan_tensor, cpu_tensor) should dispatch to the vulkan kernel
+    Dense = auto()
+
```
> This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of dispatch_key_set.py), but the new logic will definitely be slower. It currently consists of:

Is there an explanation of what happens today (how one goes from a DispatchKeySet(is this the right analogy?) to an index into the operator table)? Do we have some idea of how much additional performance overhead that will be right now?

bdhirsh replied:
> Is there an explanation of what happens today for comparison

Yep- for comparison, today is just a single `highestPriorityTypeId()` call. This change will add in a handful of bitshifting operations, which will definitely be slower. The hope is that bitshifting operations are generally very fast - probably faster than, say, an extra redispatch (which was one of the proposed alternatives originally).

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space - python changes"

This PR implements a (mostly faithful) version of the design to free up dispatch key space, in Python. I think it serves as a good place to surface feedback, because most of the real changes will involve porting this logic into C++ and dealing with failures as they show up.

I would take a look at some of the tests in `test_mock_dispatch.py` to get a better idea of the new semantics.

A few design highlights:

**DispatchKey.h Enum changes**

Previously, you could think of there as being two types of DispatchKeys in the enum:
- runtime keys
- alias keys

Now, I'd classify the DispatchKeys in this PR as one of five types:
- backend keys
- functionality keys
- per-backend functionality keys (the functionality **type**)
- per-backend functionality keys (the per-backend *instances* of those types)
- alias keys

backend key examples:
- `DispatchKey.CPUBit`, `DispatchKey.CUDABit`
- These map to one of the 12 bits at the end of the bitset
- These do NOT correspond to individual slots in the runtime operator table.

functionality key examples:
- `Batched`, `Named`, `Functionalize`, `FuncTorchGradWrapper`
- These map to one of the (64-12) bits in the front of the bitset
- These correspond to types of functionality that only have a single implementation per operator- different backends like CUDA/XLA cannot customize their implementation.
- These DO correspond to individual slots in the runtime operator table.


per-backend functionality examples (the types):
- `Dense`, `Sparse`, `Quantized`, `Autograd`
- These also map to one of the functionality bits in the front of the keyset
- These DO NOT correspond to individual slots in the runtime operator table: "Autograd" and "Sparse" don't correspond to a single kernel: they correspond to up to 12 individual kernels, one per backend.
- 
per-backend functionality examples (the instantiations):
- `CPU`, `AutogradCPU`, `QuantizedCPU`, `SparseCPU`
- These don't map to an individual bit in the bitset. Instead, they map to two individual bits: one functionality bit and one backend bit.
- These DO correspond to individual slots in the runtime operator table.


**Runtime key calculation changes**

This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of `dispatch_key_set.py`), but the new logic will definitely be slower. It currently consists of:
- 2x loads from a compile-time array
- 3x highestPriorityTypeId() calls (can probably drop this down to 2x)
- 3x `>>` (can probably drop this down to 2x)
- 3x `&`
- 1x `+`


The two loads show up for two reasons:
- per-functionality offset. Every functionality starts at a specific offset into the operator table. For example. `Undefined` is at entry 0, `Dense` is at entries 1-12, and `Autograd` is at entries 53-65. Once we calculate what the functionality bit of the bitset is, we need to figure out what the corresponding offset into the operator table is.
- per-backend mask. This is needed to handle fall through keys. Part of the existing functionality of the dispatcher is that you can register fall through kernels to a given key in the dispatcher, for zero runtime overhead. You are allowed to, for example, specify that `AutogradCPU` has a fallthrough kernel for a given operator, but `AutogradCUDA` does not. The logic involved in computing an index from a keyset includes masking the keyset against a bitset of "non-fallthrough keys". This means that we need to be able to represent the full cross product of keys efficiently as a mask. In this PR, I've done that by maintaining a separate mask for every backend as an array of DispatchKeySets.





[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255194).

albanD commented:
Should this be closed?
